master/internal/models/models.go

```diff
@@ -0,0 +1,71 @@
+package models
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/uptrace/bun"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+func ByIDTx(ctx context.Context, idb bun.IDB, id int) (model.RBModel, error) {
+	var out model.RBModel
+	err := idb.NewSelect().
+		Model(&out).
+		// XXX: a better way to do a count query as a subquery?
+		ColumnExpr(`(select count (*) from model_versions
+					where model_versions.model_id = models.id) as num_versions`).
+		Where("id = ?", id).
+		Scan(ctx)
+	if err != nil {
+		return model.RBModel{}, errors.Wrapf(err, "error selecting model(%v)", id)
+	}
+	return out, err
+}
+
+func ByID(ctx context.Context, id int) (model.RBModel, error) {
+	return ByIDTx(ctx, db.Bun, id)
+}
+
+func ByNameTx(ctx context.Context, idb bun.IDB, name string) (model.RBModel, error) {
+	var out model.RBModel
+	err := idb.NewSelect().
+		Model(&out).
+		// XXX: a better way to do a count query as a subquery?
+		ColumnExpr(`(select count (*) from model_versions
+					where model_versions.model_id = models.id) as num_versions`).
+		Where("name = ?", name).
+		Scan(ctx)
+	if err != nil {
+		return model.RBModel{}, errors.Wrapf(err, "error selecting model(%v)", name)
+	}
+	return out, err
+}
+
+func ByName(ctx context.Context, name string) (model.RBModel, error) {
+	return ByNameTx(ctx, db.Bun, name)
+}
+
+func VersionTx(
+	ctx context.Context, idb bun.IDB, modelId int, version int,
+) (model.RBModelVersion, error) {
+	var out model.RBModelVersion
+	err := idb.NewSelect().
+		Model(&out).
+		Where("modelId = ?", modelId).
+		Where("version = ?", version).
```

Comment:
This is different from what we currently do in WebUI. In our current setup `/models/Tester/versions/41/overview` , links to `model_versions.id = 41`, not the auto-incremented `model_versions.version`.

Reply:
good catch.  I think I messed up my interpretation of the original sql.

Reply:
So it should be:
```sql
	err := idb.NewSelect().
		Model(&out).
		Where("model_id = ?", modelId).
		Where("id = ?", version).
		Scan(ctx)
```
right?

Reply:
Wait... `model_version.version` is not auto-incremented, but `id` is:
```determined=# \d model_versions
                                          Table "public.model_versions"
      Column       |           Type           | Collation | Nullable |                  Default
-------------------+--------------------------+-----------+----------+--------------------------------------------
 version           | integer                  |           | not null |
 checkpoint_uuid   | uuid                     |           | not null |
 creation_time     | timestamp with time zone |           | not null |
 last_updated_time | timestamp with time zone |           |          |
 metadata          | jsonb                    |           |          |
 model_id          | integer                  |           | not null |
 id                | integer                  |           | not null | nextval('model_versions_id_seq'::regclass)
 name              | text                     |           |          |
 comment           | text                     |           |          |
 user_id           | integer                  |           |          |
 labels            | text[]                   |           |          |
 notes             | text                     |           |          |
```

Reply:
since ID is unique I think it can safely be just `Where("id = ?", version).`

Reply:
The old code looks buggy to me.

The go code was:
```go
QueryProto("get_model_version", resp.ModelVersion, parentModel.Model.Id, req.ModelVersion)
```

The sql was:
```sql
WITH mv AS (
  SELECT version, checkpoint_uuid, model_versions.id, creation_time, name, comment, metadata, labels, notes, username, last_updated_time
    FROM model_versions
    LEFT JOIN users ON users.id = model_versions.user_id
    WHERE model_id = $1 AND model_versions.id = $2
),
```

It looks like we were previously matching `model_versions.id` against `req.ModelVersion`.

Reply:
the increment is being done in `insert_model_version` but maybe should be defined by the table
`(SELECT COALESCE(max(version), 0) + 1 FROM model_versions WHERE model_id = $1)`

I'd say `req.ModelVersion` is named poorly rather than it being a bug, because in the Model page we link to ModelVersions by the id field: https://github.com/determined-ai/determined/blob/master/webui/react/src/components/Table.tsx#L167-L182 - it's a long time ago but I think I had a convo about which one we wanted to use

Reply:
wait... isn't the `id` field already an autoincrement field?  I thought that's what the `nextval('model_versions_id_seq'::regclass)` thing meant.  Why are we doing a manual "auto" increment on insert?

Reply:
> I'd say req.ModelVersion is named poorly rather than it being a bug

ok.